pycbc/inference/gelman_rubin.py

```diff
@@ -0,0 +1,174 @@
+""" This modules provides functions for evaluating the Gelman-Rubin convergence
+diagnostic statistic.
+"""
+
+import numpy
+
+def walk(chains, start, end, step):
+    """ Calculates Gelman-Rubin conervergence statistic along chains of data.
+    This function will advance along the chains and calculate the
+    statistic for each step.
+
+    Parameters
+    ----------
+    chains : iterable
+        An iterable of numpy.array instances that contain the samples
+        for each chain. Each chain has shape (nparameters, niterations).
+    start : float
+        Start index of blocks to calculate all statistics.
+    end : float
+        Last index of blocks to calculate statistics.
+    step : float
+        Step size to take for next block.
+
+    Returns
+    -------
+    starts : numpy.array
+        1-D array of start indexes of calculations.
+    ends : numpy.array
+        1-D array of end indexes of caluclations.
+    stats : numpy.array
+        Array with convergence statistic. It has
+        shape (nparameters, ncalculations).
+    """
+
+    # get number of chains, parameters, and iterations
+    chains = numpy.array(chains)
+    nchains, nparameters, niterations = chains.shape
+
+    # get end index of blocks
+    ends = numpy.arange(start, end, step)
+    stats = numpy.zeros((nparameters, len(ends)))
+
+    # get start index of blocks
+    starts = numpy.array(len(ends) * [start])
+
+    # loop over end indexes and calculate statistic
+    for i, e in enumerate(ends):
+        tmp = chains[:,:,0:e]
+        stats[:, i] = gelman_rubin(tmp)
+
+    return starts, ends, stats
+
+def gelman_rubin(chains, auto_burn_in=True):
+    """ Calculates the univariate Gelman-Rubin convergence statistic.
+
+    Parameters
+    ----------
+    chains : iterable
+        An iterable of numpy.array instances that contain the samples
+        for each chain. Each chain has shape (nparameters, niterations).
+    auto_burn_in : bool
+        If True, then only use later half of samples provided.
+
+    Returns
+    -------
+    psrf : numpy.array
+        A numpy.array of shape (nparameters) that has the point estimates of
+        the potential scale reduction factor.
+    """
+
+    # remove first have of samples
+    # this will have shape (nchains, nparameters, niterations)
+    if auto_burn_in:
+        _, _, niterations = numpy.array(chains).shape
+        chains = numpy.array([chain[:, niterations / 2 + 1:]
+                              for chain in chains])
+
+    # get number of chains, parameters, and iterations
+    chains = numpy.array(chains)
+    nchains, nparameters, niterations = chains.shape
+
+    # calculate the covariance matrix for each chain
+    # this will have shape (nchains, nparameters, nparameters)
+    chains_covs = numpy.array([numpy.cov(chain) for chain in chains])
+    if nparameters == 1:
+        chains_covs = chains_covs.reshape((nchains, 1, 1))
+
+    # calculate W
```

Comment:
It would be nice to write something about what W is. Like "calculate W, which is the within-chain variance"?